NewObjectToFactoryCreateRector

JanMikes opened the pull request:
Closes #981

JanMikes commented:
@TomasVotruba please could you check my code?

I am getting weird error with this config:
```
Symfony\Component\DependencyInjection\Exception\InvalidArgumentException : The configuration key "Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass" is unsupported for definition "Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector" in "/Users/janmikes/Sites/rector/tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml". Allowed configuration keys are "alias", "parent", "class", "shared", "synthetic", "lazy", "public", "abstract", "deprecated", "factory", "file", "arguments", "properties", "configurator", "calls", "tags", "decorates", "decoration_inner_name", "decoration_priority", "autowire", "autoconfigure", "bind".
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:813
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:341
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:229
 /Users/janmikes/Sites/rector/vendor/symplify/package-builder/src/Reflection/PrivatesCaller.php:23
 /Users/janmikes/Sites/rector/vendor/symplify/package-builder/src/Yaml/FileLoader/AbstractParameterMergingYamlFileLoader.php:106
 /Users/janmikes/Sites/rector/vendor/symfony/config/Loader/DelegatingLoader.php:40
 /Users/janmikes/Sites/rector/src/DependencyInjection/RectorKernel.php:50
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:653
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:543
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:133
 /Users/janmikes/Sites/rector/src/DependencyInjection/ContainerFactory.php:26
 /Users/janmikes/Sites/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:142
 /Users/janmikes/Sites/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:64
```

JanMikes commented:
Nevermind, i fixed that. Not using `provideConfig()` but `getRectorConfiguration()` did the trick 👍

TomasVotruba commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml:
```diff
@@ -0,0 +1,3 @@
+services:
+  Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+    Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass: []
```
Please use 4-spacing as everywhere here. It shold be loaded `.editorconfig`. Do you have it installed in PHPStorm?

JanMikes replied:
No i did not, thanks for perfect tip 👍

TomasVotruba commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/NewObjectToFactoryCreateRectorTest.php:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * @covers \Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector
+ * @covers \Rector\Rector\Architecture\DependencyInjection\ReplaceVariableByPropertyFetchRector
```
This is unneded step. Use just single Rector to do this, so people won't forget

TomasVotruba replied:
Yea, the goal of this annotaiton is to get to tested class quickly. Just control + click in PHPStorm.

In 99 % cases it's in `getRectorClass()` so it's not needed here.

JanMikes commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml:
```diff
@@ -0,0 +1,6 @@
+services:
+  Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+    arguments:
```
@TomasVotruba well, this works, but please how would the config look like, if i wanted to use more objects and more factories, because `arguments` key is unique for this service. Sorry for struggling with this basic symfony/yaml feature

TomasVotruba replied:
That's part of learning Symfony.

It took me while to find this feature, it's here: https://symfony.com/blog/new-in-symfony-3-3-simpler-service-configuration#short-syntax-for-service-configuration

TomasVotruba commented:
```php
<?php

final class NewObjectToFactoryCreateRector extends AbstractRector
{
    /**
     * @var string[][]
     */
    private $factoryWithMethodByObject;

    /**
     * @param string[][] $factoryWithMethodByObject
     */
    public function __construct(array $factoryWithMethodByObject)
    {
        $this->factoryWithMethodByObject = $factoryWithMethodByObject;
    }
    
    /**
     * @return string[]
     */
    public function getNodeTypes(): array
    {
        return [New_::class];
    }

    /**
     * @param New_ $node
     */
    public function refactor(Node $node): ?Node
    {
        foreach ($this->factoryWithMethodByObject as $type => $factoryWithMethod) {
            if (! $this->isType($node, $type)) {
                continue;
            }

            // reactor to MethodCall
            // add property
        }

        return $node;
    }
}
```

This will be then configured in config:


```yaml
# rector.yml
services:
	NewObjectToFactoryCreateRector: # ← rector class
		# configuration array that will be passed as $factoryWithMethodByObject
		FirstObject: ['FirstFactory', 'method']
		SecondObject: ['SecondFactory', 'method']
```

↓ full syntax without magic. This would people have to write manually and be careful about typos in "factoryWithMethodByObject", which can be really annoying.

```yaml
services:
	NewObjectToFactoryCreateRector: # ← rector class
	    arguments:
	    	$factoryWithMethodByObject:
				FirstObject: ['FirstFactory', 'method']
				SecondObject: ['SecondFactory', 'method']
```

TomasVotruba commented:
Could you push it? Now there is no array in constructor to put it to

TomasVotruba commented:
It works as expected. Here is the change in code: https://github.com/rectorphp/rector/compare/new-object-to-factory-method-rector...quick-example?expand=1

JanMikes commented:
Btw i did not test/cover edge cases - property already exists, factory already exists in different property in class etc.

TomasVotruba commented:
@JanMikes I see some commits already in master. Could you rebase on it?

![image](https://user-images.githubusercontent.com/924196/51703087-bc799f00-2015-11e9-8f19-376229aab6a9.png)

TomasVotruba commented:
Feature works, nice :+1: 

Testing edge cases would be nice to have, but not neccesary. It's up to you.
I think we'll find them when using on Entrydo.

Also there is 1 fail on static analysis, please fix it:

https://travis-ci.org/rectorphp/rector/jobs/483997841#L661

JanMikes commented:
Thank you, i fixed that and i will as well cover the edge cases, while i am already in it 😉 

I think that history must be mistaken, i have just checked it and there are no commits in master yet -> i rebased this branch on master. Last my commit i can see in master is  `31a5e9bcd2c44a1553fd7f17f4d4c563c138af3b`

JanMikes commented:
This edge case covered by: https://github.com/rectorphp/rector/blob/43de77013c5ca172fb53a87cb35ec13e2391c289/tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture2.php.inc

Now it is definitely ready to merge 😉

TomasVotruba commented:
I'm on it :)

TomasVotruba commented:
Btw, no need to rebase after each commit, just once in the end.
Now I don't know what's new since the last review and have to read whole code.

TomasVotruba commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc:
```diff
@@ -0,0 +1,37 @@
+{
+    public function default()
+    {
+        new MyClass('abcd');
```
Could add a new line under this one with assign case?

```php
$myClass = new MyClass('abcd');
```

JanMikes replied:
Good catch! This causes error and i have to figure out and fix it 😉 👍 

Anyways, it is terrible for debugging with xdebug, i can not see what exact function call caused it and overall i have no idea 😄 

<img width="871" alt="screenshot 2019-01-24 22 58 10" src="https://user-images.githubusercontent.com/3995003/51711170-95c66300-202b-11e9-801a-42309e910792.png">

JanMikes replied:
After quick research i found out this is the problem:

```
$this->addNodeAfterNode($methodCallNode, $node);
$this->removeNode($node);
```

Because `removeNode()` only works if it is `new xx()` without variable assign.

I will figure out and dig more and let you know 😉 im looking for elegant way of replacing New with MethodCall node.

JanMikes replied:
Simple `return $methodCallNode;` did the trick 🎉

TomasVotruba commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture2.php.inc:
```diff
@@ -0,0 +1,45 @@
+        $this->mySomeFactory->create('abcd');
+    }
+}
+?>
```
Awesome :cupid:

TomasVotruba commented on tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml:
```diff
@@ -0,0 +1,5 @@
+    Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+        Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass:
+            - Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
+            - create
```
What about adding `class` and `method` keys, so it's more obvious this is not n-list?

```diff
 services:
     Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
         Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass:
-            - Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
-            - create
+            class: Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
+            method create
```

TomasVotruba commented on src/Rector/Architecture/Factory/NewObjectToFactoryCreateRector.php:
```diff
@@ -0,0 +1,126 @@
+}
+CODE_SAMPLE
+                ,
+                ['MyClass' => ['MyClassFactory', 'create']]
```
:+1:

TomasVotruba commented:
Great! Ready to merge for me :+1:

TomasVotruba commented:
Kaboom :rocket:

JanMikes commented:
Looking forward to get this tagged 😉 for now i will use dev-master dependency.
